ogre2/src/Ogre2COMVisual.cc

```diff
@@ -0,0 +1,218 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "ignition/rendering/ogre2/Ogre2COMVisual.hh"
+#include "ignition/rendering/ogre2/Ogre2DynamicRenderable.hh"
+#include "ignition/rendering/ogre2/Ogre2Material.hh"
+
+#ifdef _MSC_VER
+  #pragma warning(push, 0)
+#endif
+#include <OgreSceneNode.h>
+#ifdef _MSC_VER
+  #pragma warning(pop)
+#endif
+
+class ignition::rendering::Ogre2COMVisualPrivate
+{
+  /// \brief Grid materal
+  public: Ogre2MaterialPtr material = nullptr;
+
+  /// \brief Lines that make the cross marking the center of mass.
+  public: std::shared_ptr<Ogre2DynamicRenderable> crossLines = nullptr;
+
+  /// \brief Sphere visual marking the center of mass
+  public: VisualPtr sphereVis = nullptr;
+};
+
+using namespace ignition;
+using namespace rendering;
+
+//////////////////////////////////////////////////
+Ogre2COMVisual::Ogre2COMVisual()
+  : dataPtr(new Ogre2COMVisualPrivate)
+{
+}
+
+//////////////////////////////////////////////////
+Ogre2COMVisual::~Ogre2COMVisual()
+{
+}
+
+//////////////////////////////////////////////////
+void Ogre2COMVisual::PreRender()
+{
+  if (this->HasParent() && this->ParentName().empty())
+    this->parentName = this->Parent()->Name();
+
+  if (this->dirtyCOMVisual &&
+      !this->ParentName().empty())
+  {
+    this->parentName = this->Parent()->Name();
+    this->CreateVisual();
+    this->dirtyCOMVisual = false;
+  }
+}
+
+//////////////////////////////////////////////////
+void Ogre2COMVisual::Init()
+{
+  BaseCOMVisual::Init();
+}
+
+//////////////////////////////////////////////////
+void Ogre2COMVisual::Destroy()
+{
+  if (this->dataPtr->sphereVis != nullptr)
+  {
+    this->dataPtr->sphereVis->Destroy();
+    this->dataPtr->sphereVis.reset();
+  }
+
+  if (this->dataPtr->crossLines)
+  {
+    this->dataPtr->crossLines->Destroy();
+    this->dataPtr->crossLines.reset();
+  }
+
+  if (this->dataPtr->material && this->Scene())
+  {
+    this->Scene()->DestroyMaterial(this->dataPtr->material);
+    this->dataPtr->material.reset();
+  }
+}
+
+//////////////////////////////////////////////////
+void Ogre2COMVisual::CreateVisual()
+{
+  if (!this->dataPtr->crossLines)
+  {
+    this->dataPtr->crossLines.reset(
+        new Ogre2DynamicRenderable(this->Scene()));
+    this->ogreNode->attachObject(this->dataPtr->crossLines->OgreObject());
+  }
+
+  if (!this->dataPtr->sphereVis)
+  {
+    this->dataPtr->sphereVis = this->Scene()->CreateVisual();
+    this->dataPtr->sphereVis->AddGeometry(this->Scene()->CreateSphere());
+    this->dataPtr->sphereVis->SetMaterial("Default/CoM");
+    this->dataPtr->sphereVis->SetInheritScale(false);
+    this->AddChild(this->dataPtr->sphereVis);
+  }
+
+  // Compute radius of sphere with density of lead and equivalent mass.
+  double sphereRadius;
+  double dLead = 11340;
+  sphereRadius = cbrt((0.75 * this->Mass()) / (IGN_PI * dLead));
```

Comment:
this calculation is duplicated in both the ogre and ogre2 code. I recommend moving it to the base class to avoid duplication

Reply:
[`360086d`](https://github.com/ignitionrobotics/ign-rendering/pull/345/commits/360086d2f169ba55bb62c56ace60cdbf978ebcbb)

Reply:
perfect!